src/optionals/queue.c

```diff
@@ -0,0 +1,241 @@
+    }
+
+    Queue *queue = AS_QUEUE(args[0]);
+    
+    return newResultSuccess(vm, queue->dq[queue->front]);
+}
+
+static Value queuePush(DictuVM *vm, int argCount, Value *args) {
+    if (argCount != 1) {
+        runtimeError(vm, "push() takes 1 argument (%d given)", argCount);
+        return EMPTY_VAL;
+    }
+
+    Queue *queue = AS_QUEUE(args[0]);
+
+    if (queue->capacity < queue->count + 1) {
+        int oldCapacity = queue->capacity;
+        queue->capacity = GROW_CAPACITY(oldCapacity);
+        queue->dq = GROW_ARRAY(vm, queue->dq, Value, oldCapacity, queue->capacity);
+    }
+
+    if (queue->rear == queue->capacity-1) {
+        queue->rear = -1;            
+    }
+
+    queue->dq[++queue->rear] = args[1];
+    queue->count++;
+
+    return NIL_VAL;
+}
+
+static Value queuePop(DictuVM *vm, int argCount, Value *args) {
+    if (argCount != 0) {
+        runtimeError(vm, "pop() takes no arguments (%d given)", argCount);
+        return EMPTY_VAL;
+    }
+
+    Queue *queue = AS_QUEUE(args[0]);
+
+    if (queue->count == 0) {
+        runtimeError(vm, "pop() called on an empty queue");
+        return EMPTY_VAL;
+    }
+
+    Value data = queue->dq[queue->front++];
+
+    if (queue->front == queue->capacity) {
+        queue->front = 0;
+    }
+
+    queue->count--;
+
+    if (queue->count < queue->capacity / 2 && queue->capacity > DEFAULT_QUEUE_CAPACITY) {
+        int oldCapacity = queue->capacity;
+        queue->capacity = SHRINK_CAPACITY(oldCapacity);
+        queue->dq = SHRINK_ARRAY(vm, queue->dq, Value, oldCapacity, queue->capacity);
+    }
+
+    return data;
+}
+
+ObjAbstract* newQueueObj(DictuVM *vm) {
+    ObjAbstract *abstract = newAbstract(vm, freeQueue, queueToString);
+    push(vm, OBJ_VAL(abstract));
+
+    Queue *queue = ALLOCATE(vm, Queue, 1);
+
+    /**
+     * Setup Queue object methods
+     */
+    defineNative(vm, &abstract->values, "isEmpty", queueIsEmpty);
+    defineNative(vm, &abstract->values, "isFull", queueIsFull);
+    defineNative(vm, &abstract->values, "cap", queueCap);
+    defineNative(vm, &abstract->values, "len", queueLen);
+    defineNative(vm, &abstract->values, "peek",  queuePeek);
+    defineNative(vm, &abstract->values, "push", queuePush);
+    defineNative(vm, &abstract->values, "pop", queuePop);
+
+    abstract->data = queue;
+    pop(vm);
+
+    return abstract;
+}
+
+static Value newQueue(DictuVM *vm, int argCount, Value *args) {
+    UNUSED(args);
+
+    if (argCount != 0) {
+        runtimeError(vm, "new() takes no arguments (%d given)", argCount);
+        return EMPTY_VAL;
+    }
+
+    ObjAbstract *abstract = newQueueObj(vm);
+    push(vm, OBJ_VAL(abstract));
+    
+    Queue *queue = abstract->data;
+
+    queue->dq = ALLOCATE(vm, Value, DEFAULT_QUEUE_CAPACITY);
+    queue->capacity = DEFAULT_QUEUE_CAPACITY;
+    queue->front = 0;
+    queue->rear = -1;
+    queue->count = 0;
+
+    Value success = newResultSuccess(vm, OBJ_VAL(abstract));
+    pop(vm);
+
+    return success;
```

Comment:
Sorry this one is on me for not being clear in the inital feedback, since we can't get an error state when creating a Queue object with `.new` we can just return the Queue directly, means we won't need to worry about unwrapping it

```suggestion
    pop(vm);

    return OBJ_VAL(abstract);
```